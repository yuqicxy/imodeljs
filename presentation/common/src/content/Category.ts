/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Content
 */

/**
 * A data structure that describes a [[Field]] category.
 * @public
 */
export interface CategoryDescription {
  /** Unique name */
  name: string;
  /** Display label */
  label: string;
  /** Extensive description */
  description: string;
  /** Priority. Categories with higher priority should appear higher in the UI */
  priority: number;
  /** Should this category be auto-expanded when it's displayed in the UI */
  expand: boolean;
}
