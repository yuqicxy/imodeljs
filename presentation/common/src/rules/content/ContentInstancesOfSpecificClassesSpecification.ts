/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module PresentationRules
 */

import { ContentSpecificationBase, ContentSpecificationTypes } from "./ContentSpecification";
import { MultiSchemaClassesSpecification } from "../ClassSpecifications";

/**
 * Creates content for ECInstance(s) of specified classes.
 *
 * **Note**: this specification doesn't rely on selection. It always returns instances no matter
 * what the selection is, so pre-filtering should be done in [[ContentRule]] condition and [[instanceFilter]].
 *
 * @public
 */
export interface ContentInstancesOfSpecificClassesSpecification extends ContentSpecificationBase {
  /** Used for serializing to JSON. */
  specType: ContentSpecificationTypes.ContentInstancesOfSpecificClasses;

  /**
   * List of ECClass specifications whose ECInstances should be selected.
   */
  classes: MultiSchemaClassesSpecification | MultiSchemaClassesSpecification[];

  /**
   * Should all [[classes]] be treated polymorphically.
   * @deprecated Use `handleInstancesPolymorphically`
   */
  arePolymorphic?: boolean;

  /**
   * Should instances be queried using a polymorphic query - from `classes` and all their
   * subclasses. This doesn't mean the resulting content will have all properties of the subclasses
   * though - they're only taken from base classes specified in `classes` attribute.
   */
  handleInstancesPolymorphically?: boolean;

  /**
   * Condition for filtering instances of defined classes.
   *
   * **See:** [ECExpressions available in instance filter]($docs/learning/presentation/Content/ECExpressions.md#instance-filter)
   */
  instanceFilter?: string;
}
