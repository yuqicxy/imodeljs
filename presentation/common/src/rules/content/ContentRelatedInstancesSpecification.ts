/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module PresentationRules
 */

import { ContentSpecificationBase, ContentSpecificationTypes } from "./ContentSpecification";
import { RelationshipDirection } from "../RelationshipDirection";
import { MultiSchemaClassesSpecification } from "../ClassSpecifications";

/**
 * Creates content for ECInstance(s) related to current selection.
 *
 * **Note:** Use [[ContentRule.condition]] to apply specification on correct selection.
 *
 * @public
 */
export interface ContentRelatedInstancesSpecification extends ContentSpecificationBase {
  /** Used for serializing to JSON. */
  specType: ContentSpecificationTypes.ContentRelatedInstances;

  /**
   * List of ECRelationship specifications to follow when looking for related instances.
   * Optional if [[relatedClasses]] is specified.
   */
  relationships?: MultiSchemaClassesSpecification | MultiSchemaClassesSpecification[];

  /**
   * List of related instance ECClass specifications. Optional if [[relationships]] is specified.
   */
  relatedClasses?: MultiSchemaClassesSpecification | MultiSchemaClassesSpecification[];

  /**
   * Direction that will be followed in the relationship select criteria.
   * Defaults to [[RelationshipDirection.Both]].
   */
  requiredDirection?: RelationshipDirection;

  /**
   * Skips defined level of related items and shows next level related items. Defaults to `0`.
   *
   * **Note:** Can't be used together with [[isRecursive]].
   *
   * @type integer
   * @minimum 0
   */
  skipRelatedLevel?: number;

  /**
   * Walks the specified relationships recursively to find related instances.
   *
   * **Note:** Can't be used together with [[skipRelatedLevel]].
   *
   * **Warning:** Using this specification has significant negative performance impact.
   */
  isRecursive?: boolean;

  /**
   * Condition for filtering instances of defined related classes.
   *
   * **See:** [ECExpressions available in instance filter]($docs/learning/presentation/Content/ECExpressions.md#instance-filter)
   */
  instanceFilter?: string;
}
