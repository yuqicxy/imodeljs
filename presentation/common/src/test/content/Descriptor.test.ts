/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
import { expect } from "chai";
import * as faker from "faker";
import {
  createRandomDescriptorJSON, createRandomDescriptor, createRandomPrimitiveField, createRandomCategory,
  createRandomECClassInfo, createRandomRelationshipPath, createRandomNestedFieldJSON, createRandomPrimitiveFieldJSON
} from "../_helpers/random";
import { Descriptor, Field, NestedContentField, StructTypeDescription, PropertyValueFormat } from "../../presentation-common";
import { DescriptorJSON, DescriptorSource } from "../../content/Descriptor";

describe("Descriptor", () => {

  describe("constructor", () => {

    it("creates Descriptor from DescriptorSource", () => {
      const source: DescriptorSource = {
        contentFlags: 9,
        displayType: faker.random.word(),
        fields: [],
        filterExpression: faker.random.words(),
        selectClasses: [],
      };
      const descriptor = new Descriptor(source);
      for (const key in source) {
        if (source.hasOwnProperty(key))
          expect((descriptor as any)[key]).to.deep.eq((source as any)[key]);
      }
    });

  });

  describe("fromJSON", () => {

    let testDescriptorJSON!: DescriptorJSON;
    beforeEach(() => {
      testDescriptorJSON = createRandomDescriptorJSON();
      testDescriptorJSON.fields.push(createRandomNestedFieldJSON());
    });

    const validateParentship = (fields: Field[], parent?: Field) => {
      fields.forEach((field) => {
        expect(field.parent).to.eq(parent);
        if (field.isNestedContentField())
          validateParentship(field.nestedFields, field);
      });
    };

    it("creates valid Descriptor from valid JSON", () => {
      const descriptor = Descriptor.fromJSON(testDescriptorJSON);
      validateParentship(descriptor!.fields);
      expect(descriptor).to.matchSnapshot();
    });

    it("creates valid Descriptor from valid serialized JSON", () => {
      const descriptor = Descriptor.fromJSON(JSON.stringify(testDescriptorJSON));
      validateParentship(descriptor!.fields);
      expect(descriptor).to.matchSnapshot();
    });

    it("skips fields that fail to deserialize", () => {
      testDescriptorJSON.fields = [createRandomPrimitiveFieldJSON(), undefined as any];
      const descriptor = Descriptor.fromJSON(testDescriptorJSON);
      expect(descriptor!.fields.length).to.eq(1);
    });

    it("returns undefined for undefined JSON", () => {
      const descriptor = Descriptor.fromJSON(undefined);
      expect(descriptor).to.be.undefined;
    });

  });

  describe("getFieldByName", () => {

    it("returns undefined when there are no fields", () => {
      const descriptor = createRandomDescriptor();
      descriptor.fields = [];
      expect(descriptor.getFieldByName("test")).to.be.undefined;
    });

    it("returns undefined when field is not found", () => {
      const descriptor = createRandomDescriptor();
      const name = descriptor.fields.map((f) => f.name).join();
      expect(descriptor.getFieldByName(name, true)).to.be.undefined;
    });

    it("returns a field", () => {
      const descriptor = createRandomDescriptor();
      const field = descriptor.fields[0];
      expect(descriptor.getFieldByName(field.name)).to.eq(field);
    });

    it("returns undefined when descriptor contains nested fields but field is not found", () => {
      const descriptor = createRandomDescriptor();
      const primitiveField = createRandomPrimitiveField();
      const descr: StructTypeDescription = {
        valueFormat: PropertyValueFormat.Struct,
        typeName: faker.random.word(),
        members: [{
          type: primitiveField.type,
          label: primitiveField.label,
          name: primitiveField.name,
        }],
      };
      const nestedField = new NestedContentField(createRandomCategory(), faker.random.word(),
        faker.random.words(), descr, faker.random.boolean(), faker.random.number(),
        createRandomECClassInfo(), createRandomRelationshipPath(), [primitiveField], undefined, faker.random.boolean());
      descriptor.fields.push(nestedField);
      expect(descriptor.getFieldByName("does not exist", true)).to.be.undefined;
    });

    it("returns a nested field", () => {
      const descriptor = createRandomDescriptor();
      const primitiveField = createRandomPrimitiveField();
      const descr: StructTypeDescription = {
        valueFormat: PropertyValueFormat.Struct,
        typeName: faker.random.word(),
        members: [{
          type: primitiveField.type,
          label: primitiveField.label,
          name: primitiveField.name,
        }],
      };
      const nestedField = new NestedContentField(createRandomCategory(), faker.random.word(),
        faker.random.words(), descr, faker.random.boolean(), faker.random.number(),
        createRandomECClassInfo(), createRandomRelationshipPath(), [primitiveField], undefined, faker.random.boolean());
      descriptor.fields.push(nestedField);
      expect(descriptor.getFieldByName(primitiveField.name, true)).to.eq(primitiveField);
    });

  });

  describe("createDescriptorOverrides", () => {

    it("creates a valid object with default parameters", () => {
      const descriptor = createRandomDescriptor();
      expect(descriptor.createDescriptorOverrides()).to.matchSnapshot();
    });

    it("creates a valid object with sorting field", () => {
      const descriptor = createRandomDescriptor();
      descriptor.sortingField = descriptor.fields[0];
      expect(descriptor.createDescriptorOverrides()).to.matchSnapshot();
    });

  });

  describe("createStrippedDescriptor", () => {

    it("creates a descriptor copy with empty fields and selectClasses member arrays", () => {
      // create original and verify it's valid for testing
      const descriptor = createRandomDescriptor();
      expect(descriptor.fields.length).to.be.above(0);
      expect(descriptor.selectClasses.length).to.be.above(0);

      // create a stripped descriptor and verify it's a different object
      // and doesn't contain stripped data
      const stripped = descriptor.createStrippedDescriptor();
      expect(stripped).to.not.eq(descriptor);
      expect(stripped.fields.length).to.eq(0);
      expect(stripped.selectClasses.length).to.eq(0);

      // verify original wasn't changed
      expect(descriptor.fields.length).to.be.above(0);
      expect(descriptor.selectClasses.length).to.be.above(0);
    });

  });

});
