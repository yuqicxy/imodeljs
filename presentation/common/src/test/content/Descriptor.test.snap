// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Descriptor createDescriptorOverrides creates a valid object with default parameters 1`] = `
Object {
  "contentFlags": 0,
  "displayType": "corrupti rem excepturi",
  "filterExpression": undefined,
  "hiddenFieldNames": Array [],
  "sortDirection": undefined,
  "sortingFieldName": undefined,
}
`;

exports[`Descriptor createDescriptorOverrides creates a valid object with sorting field 1`] = `
Object {
  "contentFlags": 0,
  "displayType": "dolor magnam ut",
  "filterExpression": undefined,
  "hiddenFieldNames": Array [],
  "sortDirection": undefined,
  "sortingFieldName": "Security",
}
`;

exports[`Descriptor fromJSON creates valid Descriptor from valid JSON 1`] = `
Object {
  "connectionId": "862f29a8-cca5-4f28-b876-93d6f3422a6e",
  "contentFlags": 0,
  "contentOptions": "car",
  "displayType": "ab vitae illum",
  "fields": Array [
    Object {
      "category": Object {
        "description": "Eveniet officia occaecati.",
        "expand": true,
        "label": "Run Granite",
        "name": "initiative",
        "priority": 57226,
      },
      "editor": undefined,
      "isReadonly": false,
      "label": "Computer haptic",
      "name": "Lead",
      "priority": 8001,
      "type": Object {
        "typeName": "blob",
        "valueFormat": "Primitive",
      },
    },
    Object {
      "category": Object {
        "description": "Magnam autem odit natus.",
        "expand": true,
        "label": "reboot grow",
        "name": "transmitter",
        "priority": 69041,
      },
      "editor": undefined,
      "isReadonly": true,
      "label": "Clothing Grocery",
      "name": "Direct",
      "priority": 64024,
      "type": Object {
        "typeName": "point",
        "valueFormat": "Primitive",
      },
    },
    Object {
      "category": Object {
        "description": "Voluptatem laborum velit.",
        "expand": false,
        "label": "Upgradable",
        "name": "paradigms",
        "priority": 45160,
      },
      "editor": Object {
        "name": "Avon",
      },
      "isReadonly": true,
      "label": "Specialist architectures Auto Loan Account",
      "name": "Hryvnia",
      "priority": 9349,
      "type": Object {
        "typeName": "bit",
        "valueFormat": "Primitive",
      },
    },
    Object {
      "autoExpand": false,
      "category": Object {
        "description": "Maxime accusamus aut eligendi voluptas natus.",
        "expand": false,
        "label": "Borders Towels",
        "name": "Cambridgeshire",
        "priority": 41890,
      },
      "contentClassInfo": Object {
        "id": "0xd0ab0000002b23",
        "label": "Australian Dollar",
        "name": "Bedfordshire",
      },
      "editor": Object {
        "name": "productize",
      },
      "isReadonly": true,
      "label": "primary Chips",
      "name": "Sausages",
      "nestedFields": Array [
        Object {
          "category": Object {
            "description": "Dolor et quasi quo voluptates voluptatibus itaque perferendis ex.",
            "expand": false,
            "label": "Libyan Dinar clicks-and-mortar Spur",
            "name": "time-frame",
            "priority": 44770,
          },
          "editor": Object {
            "name": "payment",
          },
          "isReadonly": false,
          "label": "Salad",
          "name": "Kids",
          "priority": 89699,
          "type": Object {
            "typeName": "enum",
            "valueFormat": "Primitive",
          },
        },
      ],
      "pathToPrimaryClass": Array [
        Object {
          "isForwardRelationship": true,
          "isPolymorphicRelationship": true,
          "relationshipInfo": Object {
            "id": "0xd9d000000160a5",
            "label": "reciprocal algorithm",
            "name": "Drive",
          },
          "sourceClassInfo": Object {
            "id": "0xee640000013c36",
            "label": "Unbranded Hat",
            "name": "monetize",
          },
          "targetClassInfo": Object {
            "id": "0x18444000000af60",
            "label": "best-of-breed applications",
            "name": "implement",
          },
        },
        Object {
          "isForwardRelationship": false,
          "isPolymorphicRelationship": false,
          "relationshipInfo": Object {
            "id": "0xc3bb000000d79b",
            "label": "impactful Coordinator",
            "name": "Tactics",
          },
          "sourceClassInfo": Object {
            "id": "0x9d98000000747c",
            "label": "open-source applications",
            "name": "challenge",
          },
          "targetClassInfo": Object {
            "id": "0x91690000010dd8",
            "label": "Engineer bandwidth Human",
            "name": "scalable",
          },
        },
      ],
      "priority": 56244,
      "type": Object {
        "members": Array [
          Object {
            "label": "Forks",
            "name": "IB",
            "type": Object {
              "typeName": "varchar",
              "valueFormat": "Primitive",
            },
          },
        ],
        "typeName": "PNG",
        "valueFormat": "Struct",
      },
    },
  ],
  "inputKeysHash": "1057ab70-e554-45dd-8cfe-19d5ce0e0284",
  "selectClasses": Array [
    Object {
      "isSelectPolymorphic": false,
      "navigationPropertyClasses": Array [
        Object {
          "isForwardRelationship": false,
          "isPolymorphicRelationship": true,
          "relationshipInfo": Object {
            "id": "0xe4410000010861",
            "label": "Corporate Wisconsin 1080p",
            "name": "Iraq",
          },
          "sourceClassInfo": Object {
            "id": "0x148050000017254",
            "label": "Montana 1080p Sports",
            "name": "JSON",
          },
          "targetClassInfo": Object {
            "id": "0xde5a000001461d",
            "label": "indexing white compressing",
            "name": "system-worthy",
          },
        },
      ],
      "pathToPrimaryClass": Array [
        Object {
          "isForwardRelationship": false,
          "isPolymorphicRelationship": false,
          "relationshipInfo": Object {
            "id": "0x18008000001616d",
            "label": "Centers Mouse",
            "name": "ability",
          },
          "sourceClassInfo": Object {
            "id": "0x17ff20000008039",
            "label": "drive",
            "name": "Paradigm",
          },
          "targetClassInfo": Object {
            "id": "0x13cd5000000267e",
            "label": "Cotton",
            "name": "turn-key",
          },
        },
        Object {
          "isForwardRelationship": false,
          "isPolymorphicRelationship": true,
          "relationshipInfo": Object {
            "id": "0x169410000017a8b",
            "label": "Kansas Diverse",
            "name": "fault-tolerant",
          },
          "sourceClassInfo": Object {
            "id": "0xf6e2000000ad93",
            "label": "Toys",
            "name": "Checking Account",
          },
          "targetClassInfo": Object {
            "id": "0xcc3b0000015531",
            "label": "e-commerce Forward magnetic",
            "name": "Electronics",
          },
        },
      ],
      "relatedInstanceClasses": Array [
        Object {
          "isForwardRelationship": true,
          "isPolymorphicRelationship": true,
          "relationshipInfo": Object {
            "id": "0x141a4000000d023",
            "label": "Bedfordshire",
            "name": "strategic",
          },
          "sourceClassInfo": Object {
            "id": "0x153b6000000f8dd",
            "label": "turquoise alarm",
            "name": "Developer",
          },
          "targetClassInfo": Object {
            "id": "0x12eaa000000dd90",
            "label": "Bedfordshire",
            "name": "back up",
          },
        },
      ],
      "relatedPropertyPaths": Array [
        Array [
          Object {
            "isForwardRelationship": false,
            "isPolymorphicRelationship": true,
            "relationshipInfo": Object {
              "id": "0x172c60000004caf",
              "label": "Malagasy Ariary Gorgeous",
              "name": "access",
            },
            "sourceClassInfo": Object {
              "id": "0x120540000008fb4",
              "label": "neutral",
              "name": "Qatari Rial",
            },
            "targetClassInfo": Object {
              "id": "0x10c700000009333",
              "label": "Research Games override",
              "name": "Synergistic",
            },
          },
        ],
        Array [
          Object {
            "isForwardRelationship": true,
            "isPolymorphicRelationship": true,
            "relationshipInfo": Object {
              "id": "0xff46000000b3f7",
              "label": "Michigan",
              "name": "Unbranded Metal Shoes",
            },
            "sourceClassInfo": Object {
              "id": "0xf8a00000179b8",
              "label": "Integration digital Rubber",
              "name": "Frozen",
            },
            "targetClassInfo": Object {
              "id": "0x3b0000003ef3",
              "label": "seize deploy Comoro Franc",
              "name": "Sleek",
            },
          },
        ],
      ],
      "selectClassInfo": Object {
        "id": "0xffec0000007f41",
        "label": "Cambridgeshire",
        "name": "microchip",
      },
    },
    Object {
      "isSelectPolymorphic": false,
      "navigationPropertyClasses": Array [
        Object {
          "isForwardRelationship": true,
          "isPolymorphicRelationship": false,
          "relationshipInfo": Object {
            "id": "0xbb7300000145f8",
            "label": "solid state",
            "name": "parsing",
          },
          "sourceClassInfo": Object {
            "id": "0xc0540000015b22",
            "label": "white optical",
            "name": "port",
          },
          "targetClassInfo": Object {
            "id": "0x7a65000000e109",
            "label": "zero defect Berkshire RAM",
            "name": "Legacy",
          },
        },
      ],
      "pathToPrimaryClass": Array [
        Object {
          "isForwardRelationship": true,
          "isPolymorphicRelationship": true,
          "relationshipInfo": Object {
            "id": "0xb58100000011b0",
            "label": "data-warehouse",
            "name": "Dobra",
          },
          "sourceClassInfo": Object {
            "id": "0x36180000012d5c",
            "label": "Concrete card Buckinghamshire",
            "name": "Licensed",
          },
          "targetClassInfo": Object {
            "id": "0x12aa7000000ada2",
            "label": "Steel productivity Practical",
            "name": "Refined",
          },
        },
        Object {
          "isForwardRelationship": true,
          "isPolymorphicRelationship": false,
          "relationshipInfo": Object {
            "id": "0x1524e0000004dac",
            "label": "Books Small Connecticut",
            "name": "Reverse-engineered",
          },
          "sourceClassInfo": Object {
            "id": "0xa379000000a5a1",
            "label": "Sports",
            "name": "Home Loan Account",
          },
          "targetClassInfo": Object {
            "id": "0x1713f0000003ff9",
            "label": "Chief cyan e-commerce",
            "name": "Fields",
          },
        },
      ],
      "relatedInstanceClasses": Array [
        Object {
          "isForwardRelationship": false,
          "isPolymorphicRelationship": false,
          "relationshipInfo": Object {
            "id": "0x447f000001520e",
            "label": "Personal Loan Account Jewelery",
            "name": "asynchronous",
          },
          "sourceClassInfo": Object {
            "id": "0x98fd00000185aa",
            "label": "generating",
            "name": "Ball",
          },
          "targetClassInfo": Object {
            "id": "0xea55000000a647",
            "label": "blue Agent morph",
            "name": "productize",
          },
        },
      ],
      "relatedPropertyPaths": Array [
        Array [
          Object {
            "isForwardRelationship": true,
            "isPolymorphicRelationship": false,
            "relationshipInfo": Object {
              "id": "0x158060000014c20",
              "label": "Versatile",
              "name": "blue",
            },
            "sourceClassInfo": Object {
              "id": "0x9608000000332b",
              "label": "salmon",
              "name": "next generation",
            },
            "targetClassInfo": Object {
              "id": "0xb94b00000030d6",
              "label": "Ameliorated US Dollar e-services",
              "name": "Togo",
            },
          },
        ],
        Array [
          Object {
            "isForwardRelationship": true,
            "isPolymorphicRelationship": true,
            "relationshipInfo": Object {
              "id": "0x160e6000000d7fd",
              "label": "Metal",
              "name": "withdrawal",
            },
            "sourceClassInfo": Object {
              "id": "0x78a5000000630b",
              "label": "Republic of Korea Mouse Chief",
              "name": "Borders",
            },
            "targetClassInfo": Object {
              "id": "0x117ea00000093ca",
              "label": "Human Borders Small Cotton Soap",
              "name": "bypassing",
            },
          },
        ],
      ],
      "selectClassInfo": Object {
        "id": "0x8c4500000037e6",
        "label": "Consultant software system",
        "name": "red",
      },
    },
  ],
}
`;

exports[`Descriptor fromJSON creates valid Descriptor from valid serialized JSON 1`] = `
Object {
  "connectionId": "98bd5a04-7c21-474c-8bda-3d75aad27a16",
  "contentFlags": 0,
  "contentOptions": "car",
  "displayType": "sapiente labore officia",
  "fields": Array [
    Object {
      "category": Object {
        "description": "Et accusantium rem placeat magni voluptatem voluptatem quam velit.",
        "expand": false,
        "label": "bus",
        "name": "Tasty",
        "priority": 42018,
      },
      "editor": undefined,
      "isReadonly": true,
      "label": "transmitting",
      "name": "overriding",
      "priority": 42266,
      "type": Object {
        "typeName": "point",
        "valueFormat": "Primitive",
      },
    },
    Object {
      "category": Object {
        "description": "Est omnis in repellendus repellat porro rerum optio dolorem voluptatum.",
        "expand": true,
        "label": "web-enabled Awesome digital",
        "name": "Steel",
        "priority": 46425,
      },
      "editor": undefined,
      "isReadonly": true,
      "label": "Checking Account",
      "name": "Ergonomic",
      "priority": 62102,
      "type": Object {
        "typeName": "bigint",
        "valueFormat": "Primitive",
      },
    },
    Object {
      "category": Object {
        "description": "Dolorum et temporibus quia autem ut molestiae odio fugiat est.",
        "expand": true,
        "label": "invoice index Liaison",
        "name": "revolutionize",
        "priority": 39091,
      },
      "editor": Object {
        "name": "synthesizing",
      },
      "isReadonly": true,
      "label": "zero defect Global Freeway",
      "name": "Plastic",
      "priority": 15412,
      "type": Object {
        "typeName": "binary",
        "valueFormat": "Primitive",
      },
    },
    Object {
      "autoExpand": false,
      "category": Object {
        "description": "Repellat iure consequatur cumque.",
        "expand": false,
        "label": "functionalities",
        "name": "Norwegian Krone",
        "priority": 96821,
      },
      "contentClassInfo": Object {
        "id": "0x10bb20000007fa1",
        "label": "enterprise",
        "name": "indigo",
      },
      "editor": undefined,
      "isReadonly": true,
      "label": "Engineer Kansas",
      "name": "deploy",
      "nestedFields": Array [
        Object {
          "category": Object {
            "description": "Consequatur facere magnam sint corrupti.",
            "expand": false,
            "label": "quantifying Colorado Berkshire",
            "name": "Sudan",
            "priority": 47153,
          },
          "editor": Object {
            "name": "deposit",
          },
          "isReadonly": true,
          "label": "Engineer Grocery Fall",
          "name": "protocol",
          "priority": 32728,
          "type": Object {
            "typeName": "smallint",
            "valueFormat": "Primitive",
          },
        },
      ],
      "pathToPrimaryClass": Array [
        Object {
          "isForwardRelationship": false,
          "isPolymorphicRelationship": false,
          "relationshipInfo": Object {
            "id": "0x164200000012652",
            "label": "brand Rubber",
            "name": "Engineer",
          },
          "sourceClassInfo": Object {
            "id": "0xb8b70000013b27",
            "label": "California",
            "name": "Locks",
          },
          "targetClassInfo": Object {
            "id": "0xfef10000014a4c",
            "label": "internet solution firewall",
            "name": "transition",
          },
        },
        Object {
          "isForwardRelationship": false,
          "isPolymorphicRelationship": true,
          "relationshipInfo": Object {
            "id": "0x138a70000009646",
            "label": "synthesize Investment Account",
            "name": "Personal Loan Account",
          },
          "sourceClassInfo": Object {
            "id": "0xcee50000016902",
            "label": "robust Licensed",
            "name": "Lodge",
          },
          "targetClassInfo": Object {
            "id": "0x158d20000010e25",
            "label": "Investment Account Digitized Customizable",
            "name": "calculating",
          },
        },
      ],
      "priority": 46584,
      "type": Object {
        "members": Array [
          Object {
            "label": "initiatives",
            "name": "Rustic Granite Pizza",
            "type": Object {
              "typeName": "varchar",
              "valueFormat": "Primitive",
            },
          },
        ],
        "typeName": "Division",
        "valueFormat": "Struct",
      },
    },
  ],
  "inputKeysHash": "0e02f3dd-2571-438d-92b9-2f723d9feab5",
  "selectClasses": Array [
    Object {
      "isSelectPolymorphic": false,
      "navigationPropertyClasses": Array [
        Object {
          "isForwardRelationship": true,
          "isPolymorphicRelationship": true,
          "relationshipInfo": Object {
            "id": "0x15d970000010199",
            "label": "Awesome Generic Wooden Tuna deposit",
            "name": "Switchable",
          },
          "sourceClassInfo": Object {
            "id": "0xaabe0000008a18",
            "label": "South Carolina",
            "name": "integrated",
          },
          "targetClassInfo": Object {
            "id": "0x12c000000434a",
            "label": "Berkshire",
            "name": "Avon",
          },
        },
      ],
      "pathToPrimaryClass": Array [
        Object {
          "isForwardRelationship": true,
          "isPolymorphicRelationship": true,
          "relationshipInfo": Object {
            "id": "0x85150000007a82",
            "label": "redundant Principal",
            "name": "District",
          },
          "sourceClassInfo": Object {
            "id": "0x12baa0000010999",
            "label": "Outdoors explicit bus",
            "name": "SSL",
          },
          "targetClassInfo": Object {
            "id": "0xf8290000002ce7",
            "label": "CSS hacking",
            "name": "logistical",
          },
        },
        Object {
          "isForwardRelationship": false,
          "isPolymorphicRelationship": true,
          "relationshipInfo": Object {
            "id": "0x1057e000000661c",
            "label": "Borders",
            "name": "CFA Franc BEAC",
          },
          "sourceClassInfo": Object {
            "id": "0x1d58000000d7df",
            "label": "partnerships Tasty Metal Hat",
            "name": "payment",
          },
          "targetClassInfo": Object {
            "id": "0x1c760000012e6d",
            "label": "feed",
            "name": "mission-critical",
          },
        },
      ],
      "relatedInstanceClasses": Array [
        Object {
          "isForwardRelationship": false,
          "isPolymorphicRelationship": false,
          "relationshipInfo": Object {
            "id": "0xe7ed0000012938",
            "label": "Soft",
            "name": "Frozen",
          },
          "sourceClassInfo": Object {
            "id": "0x8c4c00000168f8",
            "label": "Auto Loan Account mint green",
            "name": "Investor",
          },
          "targetClassInfo": Object {
            "id": "0x52cf00000167f5",
            "label": "optimizing lavender",
            "name": "cross-platform",
          },
        },
      ],
      "relatedPropertyPaths": Array [
        Array [
          Object {
            "isForwardRelationship": false,
            "isPolymorphicRelationship": true,
            "relationshipInfo": Object {
              "id": "0xd9f6000000d8ec",
              "label": "Fall structure",
              "name": "Gorgeous Fresh Salad",
            },
            "sourceClassInfo": Object {
              "id": "0xccbd0000009c9a",
              "label": "Customer Beauty",
              "name": "Generic Soft Chair",
            },
            "targetClassInfo": Object {
              "id": "0xca4600000000cc",
              "label": "Fresh Argentine Peso Automotive",
              "name": "Niger",
            },
          },
        ],
        Array [
          Object {
            "isForwardRelationship": true,
            "isPolymorphicRelationship": true,
            "relationshipInfo": Object {
              "id": "0xa6760000000cc1",
              "label": "invoice Handcrafted Metal Chicken Coordinator",
              "name": "Liaison",
            },
            "sourceClassInfo": Object {
              "id": "0xa44e00000061be",
              "label": "Argentina Lead tan",
              "name": "Home Loan Account",
            },
            "targetClassInfo": Object {
              "id": "0x9b490000009fbd",
              "label": "Multi-lateral hacking",
              "name": "sticky",
            },
          },
        ],
      ],
      "selectClassInfo": Object {
        "id": "0x17663000000e023",
        "label": "Security Music sticky",
        "name": "transform",
      },
    },
    Object {
      "isSelectPolymorphic": true,
      "navigationPropertyClasses": Array [
        Object {
          "isForwardRelationship": true,
          "isPolymorphicRelationship": true,
          "relationshipInfo": Object {
            "id": "0x653b0000012794",
            "label": "Managed",
            "name": "Estate",
          },
          "sourceClassInfo": Object {
            "id": "0x15b60000000b6c6",
            "label": "Intelligent Metal Towels",
            "name": "Australian Dollar",
          },
          "targetClassInfo": Object {
            "id": "0xccf000000130aa",
            "label": "ADP",
            "name": "Pennsylvania",
          },
        },
      ],
      "pathToPrimaryClass": Array [
        Object {
          "isForwardRelationship": true,
          "isPolymorphicRelationship": true,
          "relationshipInfo": Object {
            "id": "0x65380000018675",
            "label": "Profit-focused SCSI",
            "name": "Avon",
          },
          "sourceClassInfo": Object {
            "id": "0xc58a000000b0e2",
            "label": "pixel",
            "name": "parsing",
          },
          "targetClassInfo": Object {
            "id": "0x19a5000000fb5d",
            "label": "Identity",
            "name": "quantify",
          },
        },
        Object {
          "isForwardRelationship": false,
          "isPolymorphicRelationship": true,
          "relationshipInfo": Object {
            "id": "0x1059100000034f0",
            "label": "withdrawal Fresh",
            "name": "Squares",
          },
          "sourceClassInfo": Object {
            "id": "0x15d310000009a2f",
            "label": "program turn-key",
            "name": "Rubber",
          },
          "targetClassInfo": Object {
            "id": "0x1697a0000002c37",
            "label": "Jewelery expedite Seychelles",
            "name": "software",
          },
        },
      ],
      "relatedInstanceClasses": Array [
        Object {
          "isForwardRelationship": false,
          "isPolymorphicRelationship": false,
          "relationshipInfo": Object {
            "id": "0xbb3d000000bbaa",
            "label": "Soap Money Market Account Configurable",
            "name": "Investor",
          },
          "sourceClassInfo": Object {
            "id": "0xc2c30000008f8c",
            "label": "Savings Account Gibraltar Pound SQL",
            "name": "system",
          },
          "targetClassInfo": Object {
            "id": "0x187800000105d7",
            "label": "partnerships synthesize",
            "name": "alarm",
          },
        },
      ],
      "relatedPropertyPaths": Array [
        Array [
          Object {
            "isForwardRelationship": true,
            "isPolymorphicRelationship": true,
            "relationshipInfo": Object {
              "id": "0x30a300000027ae",
              "label": "plum Cloned Optimization",
              "name": "executive",
            },
            "sourceClassInfo": Object {
              "id": "0x11b0f00000132f2",
              "label": "XML deposit Global",
              "name": "deposit",
            },
            "targetClassInfo": Object {
              "id": "0x102db0000000880",
              "label": "bandwidth-monitored",
              "name": "magnetic",
            },
          },
        ],
        Array [
          Object {
            "isForwardRelationship": false,
            "isPolymorphicRelationship": true,
            "relationshipInfo": Object {
              "id": "0x81590000014d8b",
              "label": "Response Gorgeous",
              "name": "Health",
            },
            "sourceClassInfo": Object {
              "id": "0xf3d3000000007a",
              "label": "Robust",
              "name": "Singapore Dollar",
            },
            "targetClassInfo": Object {
              "id": "0x421d000000a59b",
              "label": "Home Cheese Organized",
              "name": "Squares",
            },
          },
        ],
      ],
      "selectClassInfo": Object {
        "id": "0x23d7000000e86e",
        "label": "monitor North Dakota",
        "name": "New Zealand",
      },
    },
  ],
}
`;
